x/storage/keeper/cross_app_permission.go

```diff
@@ -0,0 +1,222 @@
+	default:
+		return sdk.ExecuteResult{
+			Err: types.ErrInvalidCrossChainPackage,
+		}
+	}
+
+	if len(result.Payload) != 0 {
+		wrapPayload := types.CrossChainPackage{
+			OperationType: operationType,
+			Package:       result.Payload,
+		}
+		result.Payload = wrapPayload.MustSerialize()
+	}
+
+	return result
+}
+
+func (app *PermissionApp) handleDeletePolicyAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, ackPackage *types.DeletePolicyAckPackage) sdk.ExecuteResult {
+	return sdk.ExecuteResult{}
+}
+
+func (app *PermissionApp) handleDeletePolicyFailAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, synPackage *types.DeletePolicySynPackage) sdk.ExecuteResult {
+	return sdk.ExecuteResult{}
+}
+
+func (app *PermissionApp) handleDeletePolicySynPackage(ctx sdk.Context, deletePolicyPackage *types.DeletePolicySynPackage) sdk.ExecuteResult {
+	err := deletePolicyPackage.ValidateBasic()
+	if err != nil {
+		return sdk.ExecuteResult{
+			Payload: types.DeletePolicyAckPackage{
+				Status:    types.StatusFail,
+				ExtraData: deletePolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: err,
+		}
+	}
+
+	policy, found := app.permissionKeeper.GetPolicyByID(ctx, math.NewUintFromBigInt(deletePolicyPackage.Id))
+	if !found {
+		return sdk.ExecuteResult{
+			Payload: types.DeletePolicyAckPackage{
+				Status:    types.StatusFail,
+				ExtraData: deletePolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: types.ErrNoSuchPolicy,
+		}
+	}
+
+	_, err = app.permissionKeeper.DeletePolicy(ctx, policy.Principal, policy.ResourceType, policy.ResourceId)
+	if err != nil {
+		return sdk.ExecuteResult{
+			Payload: types.DeletePolicyAckPackage{
+				Status:    types.StatusFail,
+				ExtraData: deletePolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: err,
+		}
+	}
+
+	return sdk.ExecuteResult{
+		Payload: types.DeletePolicyAckPackage{
+			Status:    types.StatusSuccess,
+			Id:        policy.Id.BigInt(),
+			ExtraData: deletePolicyPackage.ExtraData,
+		}.MustSerialize(),
+	}
+}
+
+func (app *PermissionApp) handleCreatePolicyAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, ackPackage *types.CreatePolicyAckPackage) sdk.ExecuteResult {
+	return sdk.ExecuteResult{}
+}
+
+func (app *PermissionApp) handleCreatePolicyFailAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, ackPackage *types.CreatePolicySynPackage) sdk.ExecuteResult {
+	return sdk.ExecuteResult{}
+}
+
+func (app *PermissionApp) handleCreatePolicySynPackage(ctx sdk.Context, createPolicyPackage *types.CreatePolicySynPackage) sdk.ExecuteResult {
+	err := createPolicyPackage.ValidateBasic()
+	if err != nil {
+		return sdk.ExecuteResult{
+			Payload: types.CreatePolicyAckPackage{
+				Status:    types.StatusFail,
+				Creator:   createPolicyPackage.Operator,
+				ExtraData: createPolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: err,
+		}
+	}
+
+	var policy permtypes.Policy
+	err = json.Unmarshal(createPolicyPackage.Data, &policy)
```

Comment:
why use json instead of abi

Reply:
1. This data is passed in by the user from the frontend/client not from solidity contracts
2. This struct natively defines the serialization of json, using other methods requires additional code support
3. The additional overhead of using json serialization is not significant here

Reply:
modified To protobuf